pkg/kubelet/volumemanager/reconciler/reconstruct.go

```diff
@@ -146,12 +146,12 @@ func (rc *reconciler) updateStates(volumesNeedUpdate map[v1.UniqueVolumeName]*gl
 			continue
 		}
 		for _, volume := range gvl.podVolumes {
-			err = rc.markVolumeState(volume, operationexecutor.VolumeMounted)
+			err = rc.markVolumeState(volume, operationexecutor.VolumeMountUncertain)
```

Comment:
This will overwrite possible _certain_ mount in ASW. We've been trying not do that if possible.

I wonder if `CheckAndMarkVolumeAsUncertainViaReconstruction` / `CheckAndMarkDeviceUncertainViaReconstruction` can be used here.
@gnufied, will it break anything?

Reply:
BTW, even with this fixed, I am quite hesitant to merge it and esp. backport it to older releases. There is a reason we reworked the reconstruction completely.

Reply:
> This will overwrite possible certain mount in ASW. We've been trying not do that if possible.

This will only mark the volume that is NOT in DSW during reconstruction. So, no matter it is marked as mounted or uncertain, the reconcile will unmount the volume for us, this PR didn't change any behavior.

Reply:
Since we are handling the volume that is NOT in DSW, the situation here is much clear than what mentioned in PR https://github.com/kubernetes/kubernetes/pull/110670#:~:text=It%20adds%20volumes,in%20step%231, which aims to fix: https://github.com/kubernetes/kubernetes/issues/96635#issuecomment-1145286675 

It marked volume as uncertain by `CheckAndMarkVolumeAsUncertainViaReconstruction` / `CheckAndMarkDeviceUncertainViaReconstruction ` and set [foundDuringReconstruction](https://github.com/kubernetes/kubernetes/blob/v1.27.4/pkg/kubelet/volumemanager/cache/actual_state_of_world.go#L246) field in ASW, to avoid the volume be marked as `unmounted` again.

Reply:
The only difference is that when multiple pod volumes share the same global mount point, and only some Pods are get deleted, `MarkDeviceAsUncertain` will trigger [`MountDevice`](https://github.com/kubernetes/kubernetes/blob/v1.27.4/pkg/volume/util/operationexecutor/operation_generator.go#L637-L659) in reconcile, whereas `MarkDeviceAsMounted` will not, which causes the Pods that are not deleted using local global mount.

[When `MountDevice` fails, the state will be neither marked as `Uncertain` or `UnMounted`](https://github.com/kubernetes/kubernetes/blob/v1.27.4/pkg/volume/util/operationexecutor/operation_generator.go#L777-L798), it is ok since our target is to mount device anyway.

[When `UnmountDevice` fails, the state will be marked as `Uncertain` as well](https://github.com/kubernetes/kubernetes/blob/v1.27.4/pkg/volume/util/operationexecutor/operation_generator.go#L957-L969)

Reply:
Yeah so concern is not that volume was *not* in DSW when it was evaluated and added to ASOW, but in older reconciler, reconstruction and other operations can proceed concurrently and hence we could undo an operation that just finished.